### PR TITLE
Recursively evolve nested attrs classes

### PR DESCRIPTION
**Pull request by sscherfke**
Fixes: #634

# Pull Request Check List

This is just a friendly reminder about the most common mistakes.  Please make sure that you tick all boxes.  But please read our [contribution guide](https://www.attrs.org/en/latest/contributing.html) at least once, it will save you unnecessary review cycles!

If an item doesn't apply to your pull request, **check it anyway** to make it apparent that there's nothing left to do.  If your pull request is a documentation fix or a trivial typo, feel free to delete the whole thing.

- [x] Added **tests** for changed code.
- [x] New features have been added to our [Hypothesis testing strategy](https://github.com/python-attrs/attrs/blob/master/tests/strategies.py).
- [x] Changes or additions to public APIs are reflected in our type stubs (files ending in ``.pyi``).
    - [x] ...and used in the stub test file `tests/typing_example.py`.
- [x] Updated **documentation** for changed code.
    - [x] New functions/classes have to be added to `docs/api.rst` by hand.
    - [x] Changes to the signature of `@attr.s()` have to be added by hand too.
    - [x] Changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/markup/para.html#directive-versionadded).  Find the appropriate next version in our [``__init__.py``](https://github.com/python-attrs/attrs/blob/master/src/attr/__init__.py) file.
- [x] Documentation in `.rst` files is written using [semantic newlines](https://rhodesmill.org/brandon/2012/one-sentence-per-line/).
- [x] Changes (and possible deprecations) have news fragments in [`changelog.d`](https://github.com/python-attrs/attrs/blob/master/changelog.d).

If you have *any* questions to *any* of the points above, just **submit and ask**!  This checklist is here to *help* you, not to deter you from contributing!

### CODE REVIEW

File: changelog.d/759.change.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Let ``evolve()`` work with nested ``attrs`` classes.  #634
```
Comment (hynek) [inline changelog.d/759.change.rst:1]:
```suggestion
``attrs.evolve()`` now works recursively with nested ``attrs`` classes.
```

If you want to link 634 too, you have to created another snippet with the same content. Unfortunately that's the towncrier way.

File: src/attr/_funcs.py [hunk 2]
```diff
@@ -319,7 +319,8 @@ def evolve(inst, **changes):
     Create a new instance, based on *inst* with *changes* applied.
 
     :param inst: Instance of a class with ``attrs`` attributes.
-    :param changes: Keyword changes in the new copy.
+    :param changes: Keyword changes in the new copy.  Nested attrs classes ca
+                    be updated by passing (nested) dicts of values.
```
Comment (hynek) [inline src/attr/_funcs.py:323]:
```suggestion
    :param changes: Keyword changes in the new copy.  Nested ``attrs`` classes
       can be updated by passing (nested) dicts of values.
```

File: docs/examples.rst [hunk 3]
```diff
@@ -618,6 +618,27 @@ In Clojure that function is called `assoc <https://clojuredocs.org/clojure.core/
    >>> i1 == i2
    False
 
+This functions also works for nested ``attrs`` classes.
+You just pass a (possibly nested) dict with changes for an attribute:
```
Comment (hynek) [inline docs/examples.rst:622]:
```suggestion
Pass a (possibly nested) dictionary with changes for an attribute:
```

File: tests/test_funcs.py [hunk 4]
```diff
@@ -597,3 +597,27 @@ class C(object):
...
+            c = attr.ib(type=N2)
+            d = attr.ib(type=int)
+
+        @attr.s
+        class C(object):
+            a = attr.ib(type=N1)
+            b = attr.ib(type=int)
+
+        c1 = C(N1(N2(1), 2), 3)
+        c2 = evolve(c1, a={"c": {"e": 23}})
```
Comment (hynek) [inline tests/test_funcs.py:622]: Maybe add an example to change b too – to a literal?

File: tests/test_funcs.py [hunk 5]
```diff
@@ -597,3 +597,27 @@ class C(object):
...
+        assert c2 == C(N1(N2(23), 2), 3)
```
Comment (hynek) [inline tests/test_funcs.py:623]:
```suggestion

        assert c2 == C(N1(N2(23), 2), 3)
```

**Review by hynek — CHANGES_REQUESTED** [review]

**Comment by sscherfke** [comment]
Applied you suggestions and updated the test to include a change to `b`. :)

**Comment by hynek** [comment]
Thanks!

File: src/attr/_funcs.py [hunk 6]
```diff
@@ -337,8 +338,13 @@ def evolve(inst, **changes):
...
         attr_name = a.name  # To deal with private attributes.
         init_name = attr_name if attr_name[0] != "_" else attr_name[1:]
+        value = getattr(inst, attr_name)
         if init_name not in changes:
-            changes[init_name] = getattr(inst, attr_name)
+            # Add original value to changes
+            changes[init_name] = value
+        elif has(value):
+            # Evolve nested attrs classes
+            changes[init_name] = evolve(value, **changes[init_name])
```
Comment (RonnyPfannschmidt) [inline src/attr/_funcs.py:347]: doesn't that break the edge case of legitimately replacing a attrs instance with a dict where  permissible?

Comment (sscherfke) [reply src/attr/_funcs.py:347]: `has(value)` makes sure we only attempt recursion when the current value is an `attrs` class.

I added a test asserting that dicts can still normaly be replaced, nonetheless. :)

**Pull request merged**